src/test/scala/Ch4Spec.scala

```diff
@@ -0,0 +1,199 @@
+    }
+
+    // Composition law
+    forAll { (x: P4Data[Int], f: Int ⇒ String, g: String ⇒ Boolean) ⇒
+      p4DataEqual(
+        p4Fmap[Int, Boolean](f andThen g)(x),
+        (p4Fmap(f) andThen p4Fmap(g))(x)
+      )
+    }
+
+
+    // Problem 5
+
+    // Check equality of BOrXToB[B, X]
+    def bOrXToBEqual[B, X](
+      d1: BOrXToB[B, X], d2: BOrXToB[B, X]
+    )(implicit arbX: Arbitrary[X]) = {
+      forAll { (x: X) =>
+        d1 match {
+          case Left(b1) => d2 match {
+            case Left(b2) => b1 shouldEqual b2
+            case Right(_) => fail
+          }
+          case Right(xToB1) => d2 match {
+            case Left(_) => fail
+            case Right(xToB2) => xToB1(x) shouldEqual xToB2(x)
+          }
+        }
+      }
+    }
+
+    // Identity Law for fmapBOrXToB
+    forAll { (x: BOrXToB[Int, String]) ⇒
+      bOrXToBEqual(fmapBOrXToB[Int, Int, String](identity[Int])(x), x) }
+
+    // Composition Law for fmapBOrXToB
+    forAll { (x: BOrXToB[Int, Boolean], f: Int ⇒ String, g: String ⇒ Long) ⇒
+      bOrXToBEqual(
+        fmapBOrXToB[Int, Long, Boolean](f andThen g)(x),
+        (fmapBOrXToB[Int, String, Boolean](f) andThen fmapBOrXToB(g))(x)
+      )
+    }
+
+    // Check equality of P4Data[B]
+    def p5DataEqual[B](d1: P5Data[B], d2: P5Data[B]) = {
+      val (bOrIntToB1, bOrStrToB1) = d1
+      val (bOrIntToB2, bOrStrToB2) = d2
+
+      forAll { (i: Int) =>
+        bOrIntToB1 match {
+          case Left(b1) => bOrIntToB2 match {
+            case Left(b2) => b1 shouldEqual b2
+            case Right(_) => fail
+          }
+          case Right(intToB1) => bOrIntToB2 match {
+            case Left(_) => fail
+            case Right(intToB2) => intToB1(i) shouldEqual intToB2(i)
+          }
+        }
+      }
+
+      forAll { (s: String) =>
+        bOrStrToB1 match {
+          case Left(b1) => bOrStrToB2 match {
+            case Left(b2) => b1 shouldEqual b2
+            case Right(_) => fail
+          }
+          case Right(strToB1) => bOrStrToB2 match {
+            case Left(_) => fail
+            case Right(strToB2) => strToB1(s) shouldEqual strToB2(s)
+          }
+        }
+      }
+    }
+
+    // Identity Law
+    forAll { (x: P5Data[String]) ⇒ p5DataEqual(p5Fmap(identity[String])(x), x) }
+
+    // Composition Law
+    forAll { (x: P5Data[Boolean], f: Boolean ⇒ Int, g: Int ⇒ String) ⇒
+      p5DataEqual(
+        p5Fmap(f andThen g)(x),
+        (p5Fmap(f) andThen p5Fmap(g))(x)
+      )
+    }
+
+
+    // Problem 6
+    import io.chymyst.ch._
+
+    // Checking Param A
+    // fails
+    // def fmapResult1[A, B, C](f: A => C): Result[A, B] => Result[C, B] = implement
+    // fails
+    // def contraFmapResult1[A, B, C](f: C => A): Result[A, B] => Result[C, B] = implement
+
+    // Checking Param B
+    // works, but curry howard lib will give compile time error due to inequivalent implementations
+    // def fmapResult1[A, B, C](f: B => C): Result[A, B] => Result[A, C] = implement
```

Comment:
You can do this:

```scala
def allFmaps[A, B, C] = allOfType[(B => C) => Result[A, B] => Result[A, C]]

allFmaps[Int, String, Boolean].length should be > 0
```

Reply:
Awesome. I will change the tests for problem 6!